File: deck/rust/code/string/symbol.rs
use std::sync::OnceLock;

use crate::string::types::SymbolEntry;

/// Non-phonetic sounds carried through unchanged. The `\` escape lets a
/// literal symbol be written in talk (`\.` is a period), the same escape
/// character every other notation uses for the job.
pub fn symbols() -> &'static [SymbolEntry] {
  static CELL: OnceLock<Vec<SymbolEntry>> = OnceLock::new();

  CELL.get_or_init(|| {
    // THE SYLLABLE BREAK, WRITTEN AS IPA WRITES IT. `.` was escaped like the
    // other punctuation, which made a break in talk look nothing like the
    // `.` the source used. It needs no escape: `.` is a modifier spelling
    // (unreleased) only INSIDE a bracketed run, so a bare one outside
    // brackets can mean nothing else.
    let mut symbols: Vec<SymbolEntry> = [
      ".", "\\?", "\\!", "\\+", "\\-", "\\*", "\\@", "\\$", "\\~", "\\_", "\\^",
      "\\\\", " ",
    ]
      .into_iter()
      .map(|talk| {
        // Every escape spells the character it escapes, and the space
        // spells itself.
        let plain = talk.strip_prefix('\\').unwrap_or(talk);

        SymbolEntry {
          talk: talk.to_string(),
          ipa: plain.to_string(),
          simple: plain.to_string(),
        }
      })
      .collect();

    // THE LINK, which `normalize_ipa` used to delete. `‿` says two words
    // were spoken as one, which the writer chose to state, so it survives
    // the normalizer and needs a talk spelling to come back through.
    //
    // THE TIE IS NOT HERE. `t͡ʃ` says the letters either side of it are ONE
    // segment, which is a claim about them rather than a character standing
    // between them. Carried as a symbol it round-tripped but could not say
    // how far the binding reached, so it is folded into the `<B>` binder by
    // `bind_ties` instead. Leaving an entry here would match it first and
    // that fold would never run.
    //
    // Spelled out rather than derived, because the loop above builds each
    // entry by stripping the backslash off its own escape and this does not
    // spell the character it escapes.
    for (talk, ipa) in [("\\&", '\u{203f}')] {
      symbols.push(SymbolEntry {
        talk: talk.to_string(),
        ipa: ipa.to_string(),
        simple: ipa.to_string(),
      });
    }

    for digit in 0..=9 {
      let digit = digit.to_string();

      symbols.push(SymbolEntry {
        talk: digit.clone(),
        ipa: digit.clone(),
        simple: digit,
      });
    }

    symbols
  })
}

File: deck/typescript/code/string/symbol.ts
import type { SymbolEntry } from './type'

// Non-phonetic sounds carried through unchanged. The `\` escape lets a
// literal symbol be written in talk (`\.` is a period), the same escape
// character every other notation uses for the job.
export function buildSymbols(): SymbolEntry[] {
  const symbols: SymbolEntry[] = [
    { talk: '\\.', ipa: '.', simple: '.' },
    { talk: '\\?', ipa: '?', simple: '?' },
    { talk: '\\!', ipa: '!', simple: '!' },
    { talk: '\\+', ipa: '+', simple: '+' },
    { talk: '\\-', ipa: '-', simple: '-' },
    { talk: '\\*', ipa: '*', simple: '*' },
    { talk: '\\@', ipa: '@', simple: '@' },
    { talk: '\\$', ipa: '$', simple: '$' },
    { talk: '\\~', ipa: '~', simple: '~' },
    { talk: '\\_', ipa: '_', simple: '_' },
    { talk: '\\^', ipa: '^', simple: '^' },
    { talk: '\\\\', ipa: '\\', simple: '\\' },
    { talk: ' ', ipa: ' ', simple: ' ' },
    // THE TIE AND THE LINK, which `normalizeIpa` used to delete.
    //
    // `t͡ʃ` is one affricate where `tʃ` may be a stop meeting a fricative
    // across a boundary, and `‿` says two words were spoken as one. Both
    // state something the writer chose to state, so both survive the
    // normalizer now and both need a talk spelling to come back through.
    //
    // They are carried rather than modelled. A tie binds the two phones
    // beside it and a fuller encoding would say so in the sound itself
    // instead of standing between them. Carrying it keeps the information
    // and round-trips, which is what was actually lost.
    { talk: '\\=', ipa: '\u{0361}', simple: '\u{0361}' },
    { talk: '\\&', ipa: '\u{203f}', simple: '\u{203f}' },
  ]

  for (let n = 0; n <= 9; n++) {
    const digit = String(n)

    symbols.push({ talk: digit, ipa: digit, simple: digit })
  }

  return symbols
}
